### PR TITLE
gh-85160: Optimized singledispatchmethod access (noticeable improvement).

### PR DESCRIPTION
**Pull request by mental32**
I benchmarked using the following code:
```py
    from functools import singledispatch, singledispatchmethod
    import timeit

    class Test:
        @singledispatchmethod
        def go(self, item, arg):
            print('general')
        
        @go.register
        def _(self, item:int, arg):
            return item + arg

    @singledispatch
    def go(item, arg):
        print('general')

    @go.register
    def _(item:int, arg):
        return item + arg

    print('m', timeit.timeit('t.go(1, 1)', globals={'t': Test()}))
    print('s', timeit.timeit('go(1, 1)', globals={'go': go}))
```

And got the following output:
```
mental@felix ~/D/o/cpython (master)> ./python /tmp/fdsm.py
m 4.794188453000061
s 0.9868643390000216
mental@felix ~/D/o/cpython (master)> git checkout 40988
Switched to branch '40988'
mental@felix ~/D/o/cpython (40988) [1]> ./python /tmp/fdsm.py
m 1.3791182540007867
s 0.9882650030003788
```

I'm still curious if access could be further optimized clueless when it comes to ideas :sweat_smile: 

https://bugs.python.org/issue40988

* Issue: gh-85160

### CODE REVIEW

**Comment by CaselIT** [comment]
Thanks for taking this up!

**Comment by CaselIT** [comment]
Comment to avoid closure

**Comment by mental32** [comment]
@rhettinger sorry for bothering you but do you have enough bandwidth to review this or can you defer to someone else?

**Comment by CaselIT** [comment]
Not sure if there is anything we should do to allow this to be reviewed

**Review by MaxwellDupre — APPROVED** [review]
On my slow system more modest, but an improvement:
11.24
2.27
to
5.83
3.54
I not that this is built on 3.10.0a2+ so there maybe further speed improvements when 3.11/3.12 is released. So a worthwhile change.

**Comment by TeamSpen210** [comment]
Since this is something that’ll have observable effects on objects, should this be documented? If you’re introspecting your objects, or are say concerned about memory usage it might be surprising behaviour.

**Comment by eendebakpt** [comment]
@mental32 @AlexWaygood By accident I created an independent implementation of the singledispatchmethod improvement. In #106448 I added a performance comparison of this PR and the other one. I also added a test for the case of a class with slots (commit 4e068d4d0d7d1edfd2f2c3cd05e17a35b99985bb). Feel free to pull that one into this branch.

Both branches give a performance improvement over current main. Hopefully one of them gets accepted.

File: Lib/functools.py [hunk 1]
```diff
@@ -949,6 +953,15 @@ def _method(*args, **kwargs):
...
         _method.register = self.register
         update_wrapper(_method, self.func)
+
+        if self.attrname is not None:
+            attrname = self.attrname
+
+            try:
+                obj.__dict__[attrname] = _method
+            except AttributeError:
+                pass  # not all objects have __dict__ (e.g. class defines slots)
```
Comment (eendebakpt) [inline Lib/functools.py:963]:
```suggestion
        if self.attrname is not None:
            try:
                obj.__dict__[self.attrname] = _method
            except AttributeError:
                pass  # not all objects have __dict__ (e.g. class defines slots)
```

**Comment by mental32** [comment]
@eendebakpt thank you for continuing with this!

for any changes or merging that you would like to see done is it okay for me to say "do whatever you need to do :)" and defer to you? (including closing this PR if the other one is sufficient in perf improvements)

**Comment by AlexWaygood** [comment]
Closing in favour of #107148 (in which you have been listed as a co-author :)

**Pull request closed without merging**